### ISSUE TITLE
Reader does not throw exceptions to indicate end of data.

### ISSUE THREAD

**Issue by shrids** [pravega/pravega#2440]
**Problem description**
Exception handling in the reader needs to be improved for the following scenarios.
a. SealedStream Scenario:

`reader.readNextEvent()` will return a `null` once all the events of the sealed stream are read. `EndOfDataException` can be thrown indicating that to the user that there are no more events to be read.

b. ReaderGroup endStreamCut Scenario:

Using `ReaderGroupConfig` we can specify the upper bound upto which data can be read (endStreamCut). On reaching the endStreamCut(s) the `reader.readNextEvent()` can throw an `EndOfDataException`.

**Problem location**
EventStreamReader, SegmentInputStreamImpl

**Suggestions for an improvement**
Throw an `EndOfDataException` explicitly to indicate that there are no more events to be read.

**Comment by shrids** [comment]
One more way of handling end of stream, we can have a new api in the EventRead Interface to indicate end of stream.
```
interface EventRead {
  boolean isEvent():
  boolean isCheckpoint():
  boolean isEndOfStream():
}
```
ref: @ https://github.com/pravega/pravega/pull/2398#issuecomment-381677595

**Comment by ShwethaSNayak** [comment]
Closed vai #6673
